azurerm/cluster/node_pool/variables.tf

```diff
@@ -0,0 +1,40 @@
+variable "configuration" {
+  # type = map(object({
+  #   node_pool_name = optional(string)
+  #   vm_size        = optional(string)
+  #   node_count     = optional(string)
+
+  #   enable_auto_scaling = optional(bool)
+  #   max_count           = optional(string)
+  #   min_count           = optional(string)
+  #   eviction_policy     = optional(string)
+
+  #   max_pods     = optional(string)
+  #   os_disk_type = optional(string)
+
+  #   use_spot       = optional(bool)
+  #   max_spot_price = optional(string)
+
+  #   node_labels        = optional(map(string))
+  #   node_taints        = optional(list(string))
+  #   availability_zones = optional(list(string))
+  # }))
```

Comment:
Somehow I was not able to get this to work with the below configuration. Any advice is welcome, but as it is right now results in correct plan.

```
  main = {
    apps = {
      node_pool_name      = "main"
      enable_auto_scaling = true
      max_count           = 9
      min_count           = 3
      node_count          = 3
      vm_size             = "Standard_D2_v3"
      availability_zones  = "1,2,3"
      node_labels         = "component=main"
      os_disk_type        = "Ephemeral"
    }

    ops = {
      min_count          = 1
      node_count         = 1
      max_count          = 5
      availability_zones = ""
      use_spot = true
    }
  }

```

Reply:
The reason why I use the object for the GKE and EKS node pool modules is, that without it all values have to be strings. That makes it hard to implement e.g. labels and taints.

The thing to be aware of about the optional object attributes, and I assume this is what caused the issue for you, is that they will by default equal null. So if you use lookup on one of the attributes, it won't ever use the default value because the attribute exists but its value is null.

For attributes where you want to pass the null value through, you can therefor simply do `local.cfg["attribute_name"]` but for maps/lists that you may be using in a `for_each` or similar, you will have to default them to empty map or set. In those cases, I found it best to manually check for null with the ternary operator and then handcraft my default value. E.g. `taints = local.cfg["taints"] != null ? local.cfg["taints"] : {}` (assuming taints is a map)

You can see examples of this [here - EKS](https://github.com/kbst/terraform-kubestack/blob/master/aws/cluster/node-pool/configuration.tf#L32) or [here - GKE](https://github.com/kbst/terraform-kubestack/blob/master/google/cluster/node-pool/configuration.tf#L34).

Reply:
Thanks for the excellent pointer, now all resolved.